src/socket.js

```diff
@@ -0,0 +1,57 @@
+let DataReceiver = {
+  callbacks: new Map(),
+  onDataReceived: (aSocketId, aData) => {
+    if (!DataReceiver.callbacks.has(aSocketId)) {
+      return;
+    }
+    DataReceiver.callbacks.get(aSocketId)._updateQueue(aData);
+  },
+};
+
+browser.sockets.onDataReceived.addListener(DataReceiver.onDataReceived);
```

Comment:
We could simplify the logic here by having each `ListeningSocket` register itself with the experiments API in the constructor. Something like `browser.sockets.onDataReceived.addListener(this.onDataReceived);` Then, we could then replace the `updateQueue` method with the following:

```javascript
_onDataReceived(id, data) {
  if (id != this.id) {
    // Drop data events not relevant to this queue
    return;
  }
  this.queue.push(data);
}
```

To make it clear, we may want to update the forEach call in the experiment API with a comment that explains we're broadcasting the message to all listeners.

Reply:
I remember doing it this way so that we only register one event listener, because there is a cost associated with transporting data between the privileged and unprivileged scopes, and by having one data listener, we only pay that cost once.